### PR TITLE
improve UX for v2 Compose

### PR DESCRIPTION
**Pull request by pmeier**
The way `Compose` is currently implemented always passes the input "packed" t the individual transforms. Meaning, if one passes `Compose([...])(image, label)` individually (on "unpacked"), the transforms inside the `Compose` receive `(image, label)`, i.e. the individual inputs packed into a tuple. 

For our builtin transforms this makes no difference, since we operate on arbitrary input structures anyway. However, this becomes relevant as soon as a user wants to add a custom transform into the same pipeline. Assuming the user doesn't want to bother with our whole protocol and knows the sample structure exactly, they like want to write something like

```python
class MyTransform(nn.Module):
    def forward(self, image, label):
        ...
```

However, dropping this into a `Compose` would not work when passing `image` and `label` individually, due to the packing behavior described above. Instead the user would have to write

```python
class MyTransform(nn.Module):
    def forward(self, sample):
        image, label = sample
        ...
```

This is somewhat awkward, since the input of the individual transform no longer conforms to the input of the `Compose`.

This transform fixes this behavior. This means, if the inputs to the `Compose` are unpacked, they also have to be unpacked in the signature of the individual transform. Same is true for the other way around: if the input to the `Compose` is packed, the individual transform also needs to take a packed input.

cc @vfdev-5

### CODE REVIEW

File: test/test_transforms_v2_refactored.py [hunk 1]
```diff
@@ -1634,3 +1636,56 @@ def test_transform_negative_degrees_error(self):
...
+            or not unpack
+            and issubclass(cls, self.UnpackedInputTransform)
+            for cls in transform_clss
+        ):
+            return
+
+        transform = transforms.Compose([cls() for cls in transform_clss])
+
+        image = make_image()
+        label = torch.tensor(3)
```
Comment (NicolasHug) [inline test/test_transforms_v2_refactored.py:1684]: aren't labels pure integers when they come from a dataset (as it the case right now)?

File: test/test_transforms_v2_refactored.py [hunk 2]
```diff
@@ -1634,3 +1636,56 @@ def test_transform_negative_degrees_error(self):
...
+        def forward(self, sample):
+            image, label = sample
+            return image, label
+
+    class UnpackedInputTransform(nn.Module):
+        def forward(self, image, label):
+            return image, label
+
+    @pytest.mark.parametrize(
+        "transform_clss",
```
Comment (NicolasHug) [inline test/test_transforms_v2_refactored.py:1656]: Why not call this `transform_class` ?

File: torchvision/transforms/v2/_container.py [hunk 3]
```diff
@@ -46,10 +46,15 @@ def __init__(self, transforms: Sequence[Callable]) -> None:
         self.transforms = transforms
 
     def forward(self, *inputs: Any) -> Any:
-        sample = inputs if len(inputs) > 1 else inputs[0]
+        needs_unpacking = len(inputs) > 1
+
+        if not self.transforms:
+            return inputs if needs_unpacking else inputs[0]
```
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/transforms/v2/_container.py:52]: Depends. The old `Compose` works as a no-op in  case you don't put any transforms in. However, this doesn't sound like a valid use case. So I'm ok in putting this into the constructor.

File: test/test_transforms_v2_refactored.py [hunk 4]
```diff
@@ -1634,3 +1636,56 @@ def test_transform_negative_degrees_error(self):
...
+        ):
+            return
+
+        transform = transforms.Compose([cls() for cls in transform_clss])
+
+        image = make_image()
+        label = torch.tensor(3)
+        packed_input = (image, label)
+
+        output = transform(*packed_input if unpack else (packed_input,))
```
[the remark this answers is not in the document]
Comment (pmeier) [reply test/test_transforms_v2_refactored.py:1687]: I'll simplify.

File: test/test_transforms_v2_refactored.py (hunk 2 above)
Comment (pmeier) [reply test/test_transforms_v2_refactored.py:1656]: It would have to be `transform_classes`, since it is a list of classes. And since we use `cls` for singular, I'm usually just append an `s` to it. I'll leave it up to you.

File: torchvision/transforms/v2/_container.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/transforms/v2/_container.py]: I went with the error. LMK if you want the no-op behavior back.

File: test/test_transforms_v2_refactored.py [hunk 5]
```diff
@@ -1634,3 +1636,64 @@ def test_transform_negative_degrees_error(self):
...
+
+class TestCompose:
+    class BuiltinTransform(transforms.Transform):
+        def _transform(self, inpt, params):
+            return inpt
+
+    class PackedInputTransform(nn.Module):
+        def forward(self, sample):
+            image, label = sample
+            return image, label
```
Comment (NicolasHug) [inline test/test_transforms_v2_refactored.py:1649]: I understand that both work, but wouldn't this make more sense as

```suggestion
    class PackedInputTransform(nn.Module):
        def forward(self, sample):
            return sample
```
?

returning `image, label` works right now, but IDK if this is something we actually want to enforce support for?

Comment (NicolasHug) [reply test/test_transforms_v2_refactored.py:1649]: errrrrrrrrrrrrrrr OK lol it's the same thing (minor pref for just retuning `sample` but not blocking)

**Review by NicolasHug — APPROVED** [review]
LGTM if green, thank you!
